service/scaladsl/client/src/main/scala/com/lightbend/lagom/scaladsl/client/ServiceClient.scala

```diff
@@ -141,31 +148,162 @@ trait LagomServiceClientComponents extends TopicFactoryProvider { self: LagomCon
+ *
+ * It is important to invoke [[StandaloneLagomClientFactory#stop()]] when the application is no longer needed,
+ * as this will trigger the shutdown of the underlying [[akka.actor.ActorSystem]] and Akka Streams [[akka.stream.Materializer]]
+ * releasing all thread and connection pools in use by the clients.
+ *
+ * There is one more component that you’ll need to provide when creating a client application, that is a service locator.
+ * It is up to you what service locator you use, it could be a third party service locator, or a service locator created
+ * from static configuration.
+ *
+ * Lagom provides a number of built-in service locators, including a [[StaticServiceLocator]], a [[RoundRobinServiceLocator]]
+ * and a [[ConfigurationServiceLocator]]. The easiest way to use these is to mix in their respective Components traits.
+ *
+ * For example, here’s a client application built using the static service locator, which uses a static URI:
+ * {{{
+ * import java.net.URI
+ * import com.lightbend.lagom.scaladsl.client._
+ * import play.api.libs.ws.ahc.AhcWSComponents
+ *
+ * val clientApplication = new StandaloneLagomClientFactory("my-client")
+ *   with StaticServiceLocatorComponents
+ *   with AhcWSComponents {
+ *
+ *   override def staticServiceUri = URI.create("http://localhost:8080")
+ * }
+ * }}}
+ *
+ *
+ * @param clientName The name of the service that is consuming the Lagom service. This will impact how calls made through clients
+ *                   generated by this factory will identify themselves.
+ * @param classLoader A classloader, it will be used to create the service proxy and needs to have the API for the client in it.
+ */
+abstract class StandaloneLagomClientFactory(
+  clientName:  String,
+  classLoader: ClassLoader = classOf[StandaloneLagomClientFactory].getClassLoader
+) extends LagomClientFactory(clientName, classLoader) {
+
+  override lazy val actorSystem: ActorSystem = new ActorSystemProvider(environment, configuration).get
+  lazy val coordinatedShutdown: CoordinatedShutdown = new CoordinatedShutdownProvider(actorSystem, applicationLifecycle).get
+  override lazy val materializer: Materializer = ActorMaterializer.create(actorSystem)
+
+  /**
+   * Stop this [[LagomClientFactory]] by shutting down the internal [[akka.actor.ActorSystem]], Akka Streams [[akka.stream.Materializer]]
+   * and internal resources.
+   */
+  override def stop(): Unit = {
+
+    // we need to use the stop method from ApplicationLifecycle because the
+    // WebSocket client register a stop hook on it
+    // ideally, we should have used a CoordinatedShutdown phase for it,
+    // but we can't know if the ActorSystem is going to be shutdown together with this Factory
+    val stopped =
+      releaseInternalResources().flatMap { _ =>
+        CoordinatedShutdownSupport.asyncShutdown(actorSystem, ClientStoppedReason)
+      }(executionContext)
```

Comment:
We had refactored that previously and started to use CoordinatedShutdown but we we must stop the application lifecycle as well, because the WebSocket client is attached to it.

Reply:
Nice! Had to review a couple of times until I noticed `StandaloneLagomClientFactory` extends `LagomClientFactory`.

An alternate implementation is to register `releaseInternalResources` as a shutdown task on Coordinated Shutdown and have `stop` only invoke `CoordinatedShutdown#run`. This way, we can control _when_ `releaseInternalResources` is invoked more carefully (instead of _before_ everything).

Reply:
The problem with this is that we don't know if `CoordinatedShutdown` will be called. We may create a factory using an external `ActorSystem` and close the factory. In that case, we need to release the internal resources.